lk2nd/device/leds.c

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/* Copyright (c) 2025 Yang Xiwen <forbidden405@outlook.com> */
+
+#include <debug.h>
+
+#include <libfdt.h>
+
+#include <lk2nd/hw/gpio.h>
+
+#include "device.h"
+
+#define LK2ND_MAX_LEDS 16
+
+enum led_state {
+	LED_STATE_UNKNOWN = -1,
+	LED_STATE_ON,
+	LED_STATE_OFF,
+	LED_STATE_KEEP,
+};
+
+static int lk2nd_leds_init(const void *dtb, int node)
+{
+	int i = 0, subnode, len;
+	struct gpiol_desc desc;
+	const char *state;
+	enum led_state estate = LED_STATE_UNKNOWN;
+
+	dprintf(SPEW, " Initializing LEDs\n");
+	dprintf(SPEW, " | label | code  | gpio        |\n");
+
+	fdt_for_each_subnode(subnode, dtb, node) {
+		if (i == LK2ND_MAX_LEDS) {
+			dprintf(CRITICAL, "leds: Too many leds defined, ignoring rest\n");
+			break;
+		}
+
+		state = fdt_getprop(dtb, subnode, "default-state", &len);
+		if (len < 0 || state[len - 1] != '\0') {
+			dprintf(CRITICAL, "leds: Failed to get property \"default-state\". Broken dtb?\n");
+			continue;
+		}
+
+		if (!strcmp(state, "on"))
+			estate = LED_STATE_ON;
+		else if (!strcmp(state, "off"))
+			estate = LED_STATE_OFF;
+		else if (!strcmp(state, "keep"))
+			estate = LED_STATE_KEEP;
+
+		switch (estate) {
+		case LED_STATE_ON:
+			gpiol_get(dtb, subnode, NULL, &desc, GPIOL_FLAGS_OUT_ASSERTED);
+			break;
+		case LED_STATE_OFF:
+			gpiol_get(dtb, subnode, NULL, &desc, GPIOL_FLAGS_OUT_DEASSERTED);
+			break;
+		case LED_STATE_KEEP:
+			/* Simply leave this pin alone */
+			break;
+		case LED_STATE_UNKNOWN:
+			dprintf(CRITICAL, "leds: Unknown \"default-state\" value: %s\n", state);
+			continue;
+		}
+
+		i++;
+
+		dprintf(SPEW, " | %5s | 0x%02x 0x%04x |\n",
```

Comment:
I'd also put the "state" here I guess but for such a simple driver that doesn't actually store anything, could as well just omit the state table completely. Line below would say "x leds configured" and that would be enough IMO.